testsuite/testsuite-microprofile-jwt/src/test/java/org/wildfly/swarm/microprofile/jwtauth/PrincipalLeakTest.java

```diff
@@ -0,0 +1,79 @@
+/**
+ * Copyright 2018 Red Hat, Inc, and individual contributors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.wildfly.swarm.microprofile.jwtauth;
+
+import org.apache.http.client.fluent.Content;
+import org.apache.http.client.fluent.Request;
+import org.jboss.arquillian.container.test.api.Deployment;
+import org.jboss.arquillian.container.test.api.RunAsClient;
+import org.jboss.arquillian.junit.Arquillian;
+import org.jboss.shrinkwrap.api.Archive;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.ClassLoaderAsset;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.wildfly.swarm.microprofile.jwtauth.roles.TestApplication;
+import org.wildfly.swarm.microprofile.jwtauth.utils.TokenUtils;
+import org.wildfly.swarm.undertow.WARArchive;
+
+import static org.fest.assertions.Assertions.assertThat;
+import static org.wildfly.swarm.microprofile.jwtauth.utils.TokenUtils.createToken;
+
+/**
+ * @author Michal Szynkiewicz, michal.l.szynkiewicz@gmail.com
+ * <br>
+ * Date: 8/8/18
+ */
+@RunWith(Arquillian.class)
+public class PrincipalLeakTest {
+    @Deployment
+    public static Archive<?> createDeployment() {
+        return initDeployment().addAsResource("project-no-roles-props.yml", "project-defaults.yml");
+    }
+
+    protected static WARArchive initDeployment() {
+        WARArchive deployment = ShrinkWrap.create(WARArchive.class);
+        deployment.addClass(SubjectExposingResource.class);
+        deployment.addClass(TestApplication.class);
+        deployment.addAsManifestResource(new ClassLoaderAsset("keys/public-key.pem"), "/MP-JWT-SIGNER");
+        return deployment;
+    }
+
+    @RunAsClient
+    @Test
+    public void subjectShouldNotLeakToNonSecuredRequest() throws Exception {
+        long start = System.currentTimeMillis();
+        try {
+            // By default, there are 64 threads/tasks
```

Comment:
is there a way to limit the number of tasks/threads for a test?
If we limit it to 1 we have a faster and more reliable test...

Reply:
Single call should be enough, though it is up to you...

Reply:
The problem the PR is fixing is that the principal is kept on thread local and it was not cleared.
So if some thread handled an authenticated request, the principal has been set for it. And then, if the same thread handled an unauthenticated request, the principal was still accessible.

Single thread is not enough to reproduce the problem because thorntail/undertow has a thread pool for handling the requests and by default the pool size is 64

Reply:
I'm not even sure if it's hardcoded to 64, the number might be derived from number of CPU cores or something. You should be able to change the number of threads by configuring the default worker pool in the `io` subsystem.

Reply:
@Ladicek thanks! I was looking at undertow configuration and didn't find anything, IO has `swarm.io.workers.KEY.task-max-threads` :)